### ISSUE TITLE
Sandbox builds

### ISSUE THREAD

**Issue by junghans** [LLNL/spack#5764]
I am really nervus about the fact that I could easily sneak a `rm -rf ~` into a package.
 
So sandbox the builds, somehow along the lines of:
```
. share/spack/setup-env.sh
spack install sandbox
SANDBOX_INTRACTV=1
. <path/to/sandbox>/usr/share/sandbox/sandbox.bashrc
addread /
addwrite ${HOME}/.spack ${HOME}/spack/opt/spack/.spack-db
spack install <malicious_package>
``` 
would be nice.

`addread` / `addwrite` are just very simple function like:
```

	adddeny()    { export SANDBOX_DENY=${SANDBOX_DENY}:$1 ; }
	addpredict() { export SANDBOX_PREDICT=${SANDBOX_PREDICT}:$1 ; }
	addread()    { export SANDBOX_READ=${SANDBOX_READ}:$1 ; }
	addwrite()   { export SANDBOX_WRITE=${SANDBOX_WRITE}:$1 ; }
	sandboxon()  { export SANDBOX_ON="1" ; }
	sandboxoff() { export SANDBOX_ON="0" ; }
```
so we could easily add that stuff inside the package class.

Other than that we only need:
```
export LD_PRELOAD="/path/to/lib/sandbox"
```

**Comment by tgamblin** [comment]
Does `sandbox` use OS-native sandboxing capabilities, like user namespaces on Linux or the macos app sandbox?

**Comment by junghans** [comment]
It is based on `LD_PRELOAD`, so it would even work on old Linux system without namespaces!

**Comment by junghans** [comment]
Actually it uses namespace if enabled!

But no support for OSX!

**Comment by junghans** [comment]
Oh well apparently it can use OSX's seatbelt!

**Comment by tgamblin** [comment]
ok nice so it's a general interface to all the stuff we'd want to use?  yay!  I'm for this.

**Comment by tgamblin** [comment]
Bazel does some of this as well -- using many of the same mechanisms. https://blog.bazel.build/2015/09/11/sandboxing.html

**Comment by junghans** [comment]
Is used by Gentoo's portage by default, so it is very well tested.

**Comment by tgamblin** [comment]
Does it work on windows?  Asking for a friend.

**Comment by junghans** [comment]
Windows, of course not ;-)

**Comment by tgamblin** [comment]
I guess when we support windows we could either add support for windows sandboxing to sandbox, or we could do something else...

**Comment by junghans** [comment]
Concerning OSX, I was mistaken, portage maps the `sandbox` calls to Apple's native thing, `sandbox-exec`, so I guess at the end of the day, we need a sandbox class. When looking at the portage code I realized one could do similar things with `fakeroot`.

**Comment by tgamblin** [comment]
By "sandbox class" do you mean that we could do this entirely within Spack?

I think "seatbelt" is another name for "sandbox" in the OS X docs.

**Comment by junghans** [comment]
Yeah, portage has a `spawn_sandbox` function, which basically prefixed every command with `sandbox`, `fakeroot` or `sandbox-exec` (and a couple of options). Plus, it exports `SANDBOX_*` for `sandbox` or make a profile for `sandbox-exec`.

I think, we could do something similar in `util.Executable()` and `setupEnvironment()`.

**Comment by tgamblin** [comment]
Why wouldn't we just run the forked build process in the sandbox?  That would catch all calls, not just those through `spack.util.Executable`.  Do we need to also do an `LD_PRELOAD` for exec's *within* the sandboxed process?

**Comment by trivialfis** [comment]
Hi, would it be possible to have sandboxed environment command too?  I'm looking for something similar to guix's `environment --container` command.
